File: leetcode/leetcode/024.cpp
//
//  024.cpp
//  leetcode
//
//  Created by Hang Zhang on 4/30/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
using namespace std;
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

void printList(ListNode *h) {
    while (h) {
        cout << h->val << " -> ";
        h = h->next;
    }
    cout << "(null)\n";
}

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        printList(head);
        
        ListNode *h = new ListNode(0);
        ListNode *p = h;
        while (head && head->next) {
            ListNode *q = head->next;
            p->next = q;
            head->next = q->next;
            q->next = head;
            
            p = head;
            head = head->next;
        }
        p->next = head;
        printList(h->next);
        return h->next;
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 024 Swap Nodes in Pairs, C++ ...\n\n";
    
    ListNode *a1 = new ListNode(1);
    ListNode *a2 = new ListNode(2);
    ListNode *a3 = new ListNode(3);
    ListNode *a4 = new ListNode(4);
    a1->next = a2;
    a2->next = a3;
    a3->next = a4;
    
    Solution sol;
    sol.swapPairs(a1);
    
    return 0;
}

File: python/024.py
#   024
#   Swap Nodes in Pairs
#   2015-04-30
#####################################################
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def swapPairs(self, head):
        printList(head)
        h = ListNode(0)
        p = h
        while head and head.next:
            q = head.next
            p.next = q
            head.next = q.next
            q.next = head
            p = head
            head = head.next
        p.next = head
        printList(h.next)
        return h.next

def printList(l):
    while l:
        print(l.val, end="->")
        l = l.next
    print("(null)")

#####################################################

a1 = ListNode(1)
a2 = ListNode(2)
a3 = ListNode(3)
a4 = ListNode(4)
a1.next = a2
a2.next = a3
a3.next = a4

sol = Solution()
sol.swapPairs(a1)

